uber/models.py

```diff
@@ -1625,6 +1817,464 @@ def must_contact(self):
+            Valid values are:
+
+            * 0 `FIXED_DISCOUNT`: `discount` is interpreted as a fixed
+                dollar amount by which the badge price should be reduced. If
+                `discount` is 49 and the badge price is normally $100, then
+                the discounted badge price would be $51.
+
+            * 1 `FIXED_PRICE`: `discount` is interpreted as the actual badge
+                price. If `discount` is 49, then the discounted badge price
+                would be $49.
+
+            * 2 `PERCENT_DISCOUNT`: `discount` is interpreted as a percentage
+                by which the badge price should be reduced. If `discount` is
+                20 and the badge price is normally $50, then the discounted
+                badge price would $40 ($50 reduced by 20%). If `discount` is
+                100, then the price would be 100% off, i.e. a free badge.
+
+        expiration_date (datetime): The date & time upon which this promo code
+            expires. An expired promo code may no longer be used to receive
+            discounted badges.
+        is_expired (bool): True if this promo code is expired, False otherwise.
+        is_unlimited (bool): True if this promo code may be used an unlimited
+            number of times, False otherwise.
+        is_valid (bool): True if this promo code is still valid and may be
+            used again, False otherwise.
+        normalized_code (str): A normalized version of `code` suitable for
+            database queries. Normalization converts `code` to all lowercase
+            and removes dashes ("-").
+        used_by (list): List of attendees that have used this promo code.
+            Note:
+                This property is declared as a backref in the Attendee class.
+        uses_allowed (int): The total number of times this promo code may be
+            used. A value of None means this promo code may be used an
+            unlimited number of times.
+        uses_allowed_display (int): A human readable description of
+            uses_allowed.
+        uses_count (int): The number of times this promo code has already
+            been used.
+        uses_count_display (int): A human readable description of uses_count.
+        uses_remaining (int): Remaining number of times this promo code may
+            be used.
+        uses_remaining_display (int): A human readable description of
+            uses_remaining.
+    """
+
+    FIXED_DISCOUNT = 0
+    FIXED_PRICE = 1
+    PERCENT_DISCOUNT = 2
+    DISCOUNT_OPTS = [
+        (FIXED_DISCOUNT, 'Fixed Discount'),
+        (FIXED_PRICE, 'Fixed Price'),
+        (PERCENT_DISCOUNT, 'Percent Discount')]
+
+    AMBIGUOUS_CHARS = {
+        '0': 'OQD',
+        '1': 'IL',
+        '2': 'Z',
+        '5': 'S',
+        '6': 'G',
+        '8': 'B'}
+
+    UNAMBIGUOUS_CHARS = string.digits + string.ascii_uppercase
+    for _, s in AMBIGUOUS_CHARS.items():
```

Comment:
Huh. I'm not... super sure I support this substitution. It'd work, but I think it would still be frustrating for attendees.

Like, imagine that someone wrote down a promo code with a missing letter. The code is "LOM-105" and they wrote down "LOM-10." _We_ know that the attendee can type "L0MIO5" and we'd still accept it, but they don't, and when we reject it because they're missing the '5' they're going to sit there and spend ten minutes trying different combinations of ambiguous characters, all while being super frustrated.

I still think the normalization is good practice (since admins will enter their own promo codes and probably won't care about ambiguous characters), but for the generation of `UNAMBIGUOUS_CHARS` can we simply remove them instead of subbing in different ambiguous characters?